File: 542-01-matrix/01-matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int n = mat.size();
        int m = mat[0].size();
        vector<vector<int>> vis(n,vector<int>(m,0));
        vector<vector<int>> res(n,vector<int>(m,0));
        queue<pair<pair<int,int>,int>>q;
        for(int i =0;i<n;i++){
            for(int j=0;j<m;j++){
                if (mat[i][j] == 0) {
                    vis[i][j] = 1;
                    q.push({{i, j}, 0});}
            }
        }
        int drow[] = {-1, 0, 1, 0};
        int dcol[] = {0, 1, 0, -1};

        while(!q.empty()){
            int r = q.front().first.first;
            int c = q.front().first.second;
            int s = q.front().second;
            q.pop();
            

            for(int i =0;i<4;i++){
                int n_r = r + drow[i];
                int n_c = c + dcol[i];
                if(n_r>=0 && n_r<n && n_c>=0 && n_c<m && vis[n_r][n_c] == 0){
                    vis[n_r][n_c] = 1;
                    res[n_r][n_c] = s+1;
                    q.push({{n_r,n_c},s+1});
                }
            }
        }
        return res;
    }
};

File: 542-01-matrix/01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        n, m = len(mat), len(mat[0])
        res = [[0] * m for _ in range(n)]
        vis = [[0] * m for _ in range(n)]
        q = deque()
        for i in range(n):
            for j in range(m):
                if mat[i][j] == 0:
                    vis[i][j] = 1
                    q.append(((i, j), 0))

        drow = [-1, 0, 1, 0]
        dcol = [0, 1, 0, -1]

        while q:
            (r, c), s = q.popleft()
            res[r][c] = s

            for i in range(4):
                nr, nc = r + drow[i], c + dcol[i]
                if 0 <= nr < n and 0 <= nc < m and vis[nr][nc] == 0:
                    vis[nr][nc] = 1
                    q.append(((nr, nc), s + 1))

        return res